arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Bracket.kt

```diff
@@ -66,22 +77,143 @@ suspend fun <A> onCancel(
+ * Prefer [Resource] or [bracketCase] for the acquisition and release of resources.
+ *
+ * @param fa program that you want to register handler on
+ * @param finalizer handler to run after [fa].
+ * @see guarantee for registering a handler that ignores the [ExitCase] of [fa].
+ */
 suspend fun <A> guaranteeCase(
   fa: suspend () -> A,
-  release: suspend (ExitCase) -> Unit
-): A = bracketCase({ Unit }, { fa.invoke() }, { _, ex -> release(ex) })
+  finalizer: suspend (ExitCase) -> Unit
+): A = bracketCase({ Unit }, { fa.invoke() }, { _, ex -> finalizer(ex) })
 
+/**
+ * Meant for specifying tasks with safe resource acquisition and release in the face of errors and interruption.
+ * It would be the the equivalent of `try/catch/finally` statements in mainstream imperative languages for resource
+ * acquisition and release.
+ *
+ * @param acquire the action to acquire the resource
+ *
+ * @param use is the action to consume the resource and produce a result.
+ * Once the resulting suspend program terminates, either successfully, error or disposed,
+ * the [release] function will run to clean up the resources.
+ *
+ * @param release is the action that's supposed to release the allocated resource after `use` is done, irregardless
+ * of its exit condition.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.IO
+ *
+ * class File(url: String) {
+ *   fun open(): File = this
+ *   fun close(): Unit {}
+ *   override fun toString(): String = "This file contains some interesting content!"
+ * }
+ *
+ * suspend fun openFile(uri: String): File = File(uri).open()
+ * suspend fun closeFile(file: File): Unit = file.close()
+ * suspend fun fileToString(file: File): String = file.toString()
+ *
+ * suspend fun main(): Unit {
+ *   //sampleStart
+ *   val res = bracketCase(
+ *     acquire = { openFile("data.json") },
+ *     use = { file -> fileToString(file) }),
+ *     release = { file: File -> closeFile(file) }
+ *   )
+ *   //sampleEnd
+ *   println(res)
+ * }
+ * ```
+ */
 suspend fun <A, B> bracket(
   acquire: suspend () -> A,
   use: suspend (A) -> B,
   release: suspend (A) -> Unit
 ): B = bracketCase(acquire, use, { a, _ -> release(a) })
 
+/**
+ * A way to safely acquire a resource and release in the face of errors and cancellation.
+ * It uses [ExitCase] to distinguish between different exit cases when releasing the acquired resource.
+ *
+ * [bracketCase] exists out of a three stages:
+ *   1. acquisition
+ *   2. consumption
+ *   3. releasing
+ *
+ * 1. Resource acquisition is **NON CANCELLABLE**.
+ *   If resource acquisition fails, meaning no resource was actually successfully acquired then we short-circuit the effect.
+ *   Reason being, we cannot [release] what we did not `acquire` first. Same reason we cannot call [use].
+ *   If it is successful we pass the result to stage 2 [use].
+ *
+ * 2. Resource consumption is like any other `suspend` effect. The key difference here is that it's wired in such a way that
```

Comment:
Excellent docs and clarifications!

Reply:
Should we repeat this in `bracket`, I am unsure how we can improve the documentation of these top-level functions.

> API docs are a very low profile and have little to no styling. I'm not sure how or if we need to improve our API docs, or if we should promote certain functions in a different manner. parMapN, parTupledN, raceN.